File: server/routes/roadmaps.js
// routes/roadmaps.js
const express = require('express');
const router = express.Router();
const Roadmap = require('../models/Roadmap');

// GET all roadmaps
router.get('/', async (req, res) => {
  try {
    const roadmaps = await Roadmap.find();
    res.json(roadmaps);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// GET a single roadmap by ID
router.get('/:id', async (req, res) => {
  try {
    const roadmap = await Roadmap.findById(req.params.id);
    if (!roadmap) return res.status(404).json({ message: 'Roadmap not found' });
    res.json(roadmap);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

// POST a new roadmap
router.post('/', async (req, res) => {
  const roadmap = new Roadmap(req.body);
  try {
    const newRoadmap = await roadmap.save();
    res.status(201).json(newRoadmap);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// PUT update a roadmap
router.put('/:id', async (req, res) => {
  try {
    const updatedRoadmap = await Roadmap.findByIdAndUpdate(req.params.id, req.body, { new: true });
    if (!updatedRoadmap) return res.status(404).json({ message: 'Roadmap not found' });
    res.json(updatedRoadmap);
  } catch (err) {
    res.status(400).json({ message: err.message });
  }
});

// DELETE a roadmap
router.delete('/:id', async (req, res) => {
  try {
    const deletedRoadmap = await Roadmap.findByIdAndDelete(req.params.id);
    if (!deletedRoadmap) return res.status(404).json({ message: 'Roadmap not found' });
    res.json({ message: 'Roadmap deleted successfully' });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
});

module.exports = router;

File: server/routes/roadmaps.py
# blueprints/roadmaps.py
from flask import Blueprint, request, jsonify
from models.roadmap import Roadmap, serialize_roadmap
from bson import ObjectId


roadmaps = Blueprint('roadmaps', __name__)

@roadmaps.route('/', methods=['GET'])
def get_roadmaps():
    try:
        roadmap_list = Roadmap.objects()
        data = [serialize_roadmap(r) for r in roadmap_list]
        return jsonify(data), 200
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@roadmaps.route('/<string:roadmap_id>', methods=['GET'])
def get_roadmap(roadmap_id):
    try:
        # Convert string ID to ObjectId
        roadmap = Roadmap.objects.get(id=ObjectId(roadmap_id))
        return jsonify(serialize_roadmap(roadmap)), 200
    except Roadmap.DoesNotExist:
        return jsonify({'message': 'Roadmap not found'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 500

@roadmaps.route('/', methods=['POST'])
def create_roadmap():
    data = request.get_json()
    try:
        roadmap = Roadmap(**data).save()
        return jsonify(serialize_roadmap(roadmap)), 201
    except Exception as e:
        return jsonify({'message': str(e)}), 400

@roadmaps.route('/<string:roadmap_id>', methods=['PUT'])
def update_roadmap(roadmap_id):
    data = request.get_json()
    try:
        roadmap = Roadmap.objects.get(id=roadmap_id)
        roadmap.update(**data)
        roadmap.reload()
        return jsonify(serialize_roadmap(roadmap)), 200
    except Roadmap.DoesNotExist:
        return jsonify({'message': 'Roadmap not found'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 400

@roadmaps.route('/<string:roadmap_id>', methods=['DELETE'])
def delete_roadmap(roadmap_id):
    try:
        roadmap = Roadmap.objects.get(id=roadmap_id)
        roadmap.delete()
        return jsonify({'message': 'Roadmap deleted successfully'}), 200
    except Roadmap.DoesNotExist:
        return jsonify({'message': 'Roadmap not found'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 500